### ISSUE TITLE
pytorch gradient attack 

### ISSUE THREAD

**Issue by dongshuhao** [bethgelab/foolbox#91]
https://github.com/bethgelab/foolbox/blob/021f02a10805c3a0e95cfba8c0ceeec5aae0fc5a/foolbox/models/pytorch.py#L114

Why do we need to have the grad divided by the std? I do not think it is necessary.

**Comment by wielandbrendel** [comment]
self._process_gradient simply backprops the gradient through the preprocessing step (subtraction of the mean, scaling).

**Comment by dongshuhao** [comment]
https://github.com/bethgelab/foolbox/blob/021f02a10805c3a0e95cfba8c0ceeec5aae0fc5a/foolbox/models/pytorch.py#L99
This line is for backprops. The point it is why do we need: https://github.com/bethgelab/foolbox/blob/021f02a10805c3a0e95cfba8c0ceeec5aae0fc5a/foolbox/models/base.py#L75

Based on: https://github.com/bethgelab/foolbox/blob/021f02a10805c3a0e95cfba8c0ceeec5aae0fc5a/foolbox/attacks/gradient.py#L30

The estimated grad is applied to the original image. Why do we need to do the preprocessing then?

**Comment by wielandbrendel** [comment]
See, the processing pipeline is this: original image -> preprocessed image -> model output. Most importantly, the model only sees the preprocessed image! Hence, loss.backward() gives you the gradient with respect to the **preprocessed image**. self._process_gradient takes this gradient and returns the gradient with respect to the **original image**. If there is no preprocessing then the two are identical.

**Comment by dongshuhao** [comment]
@wielandbrendel Yes, the model only do the forward on the preprocessed image (i.e., (org_img - mean) / std) and the estimated grad is wrt the preprocessed image. I got your point. Thanks! However, I think we need to reverse the preprocessing step, right? Then, why did you divide the gradient by the std again instead of times the std? In addition, can the linear reverse preprocess (e.g, div or multiply) deal with the nonlinear nature of the model correctly?

**Comment by dongshuhao** [comment]
@wielandbrendel I got it! Thanks!
